publisher.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2019 Dgraph Labs, Inc. and Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package badger
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/dgraph-io/badger/pb"
+	"github.com/dgraph-io/badger/y"
+)
+
+type subscriber struct {
+	prefixes  [][]byte
+	sendCh    chan<- *pb.KVList
+	subCloser *y.Closer
+}
+
+type publisher struct {
+	sync.Mutex
+	pubCh       chan requests
+	subscribers map[uint64]subscriber
+	nextID      uint64
+}
+
+func newPublisher() *publisher {
+	return &publisher{
+		pubCh:       make(chan requests, 10000),
+		subscribers: make(map[uint64]subscriber),
+		nextID:      0,
+	}
+}
+
+func (p *publisher) listenForUpdates(c *y.Closer) {
+	defer func() {
+		p.cleanSubscribers()
+		c.Done()
+	}()
+	for {
+		select {
+		case <-c.HasBeenClosed():
+			return
+		case reqs := <-p.pubCh:
+			p.publishUpdates(reqs)
+		}
+	}
+}
+
+func (p *publisher) publishUpdates(reqs requests) {
+	p.Lock()
+	defer func() {
+		p.Unlock()
+
+		// release all the request
+		reqs.DecrRef()
+	}()
+	for _, s := range p.subscribers {
```

Comment:
@manishrjain will it make sense to keep some kind of Trie to directly figure out the list of subscribers a req should go to. We can keep the Trie divided based on parts of prefixes, compared to doing it byte by byte. @sch00lb0y may be for now, add a todo here mentioning that we can find a way to quickly identify the subscribers for a prefix.